File: web/WebFirstLab/AjaxFree/php/CachedResultable.php
<?php
    require_once('Resultable.php');

    class CachedResultable implements Resultable {
        private $cache;
        private $resultable;

        public function __construct(Resultable $resultable) {
            $this->resultable = $resultable;
        }

        public function result() {
            if(empty($this->cache)) {
                $this->cache = $this->resultable->result();
            }

            return $this->cache;
        }
    }
?>

File: web/WebThirdLab/src/main/java/Entities/CachedResultable.java
package Entities;

import java.util.Optional;

public class CachedResultable<T> implements Resultable<T> {
    private final Resultable<T> resultable;
    private T cache;

    public CachedResultable(Resultable<T> resultable) {
        this.resultable = resultable;
    }

    @Override
    public T result() {
        return Optional.ofNullable(cache).orElseGet(() -> {
            this.cache = resultable.result();

            return this.cache;
        });
    }
}
